### PR TITLE
[wasm] Modify HttpClient to use a default handler provider.

### PR DESCRIPTION
**Pull request by kjpou1**
- Add a private static variable to HttpClient called `Func<HttpMessageHandler> GetHttpMessageHandler` that will be used to obtain the default message handler and can be set for example as follows:

```
		var httpMessageHandler = typeof(HttpClient).GetField("GetHttpMessageHandler", 
                            BindingFlags.Static | 
                            BindingFlags.NonPublic);

                httpMessageHandler.SetValue(null, (Func<HttpClientHandler>) (() => {
			return new MyHttpClientHandler ();
		}));
```

- If a handler is not set or returns null the default implementation HttpClientHandler will be returned.

- Add HttpClient.wasm.cs implementation to handle the default HttpClient
- Add simple test to the wasm test harness bindings-test.
- Reference https://github.com/mono/mono/issues/9941

### CODE REVIEW

File: mcs/class/System.Net.Http/Makefile [hunk 1]
```diff
@@ -10,6 +10,9 @@ LIB_MCS_FLAGS = $(EXTRA_LIB_MCS_FLAGS)
 ifeq (monodroid,$(PROFILE))
 LIB_MCS_FLAGS += -d:XAMARIN_MODERN
 endif
+ifeq (wasm,$(PROFILE))
```
Comment (kumpera) [inline mcs/class/System.Net.Http/Makefile]: This should not be needed, we add `-d:WASM` in `mcs/build/profiles/wasm.make`.

File: mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+using System;
+using System.Reflection;
+
+namespace System.Net.Http {
+	public partial class HttpClient {
+
+        public static Func<HttpMessageHandler> GetHttpMessageHandler;
```
Comment (kumpera) [inline mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: This should be private. We cannot add public members to BCL types.

File: sdks/wasm/test.js [hunk 3]
```diff
@@ -73,7 +73,7 @@ var Module = {
 	},
 };
 
-var assemblies = [ "mscorlib.dll", "System.dll", "System.Core.dll", "Mono.Security.dll", "main.exe", "nunitlite.dll", "mini_tests.dll", "wasm_corlib_test.dll", "wasm_System_test.dll", "wasm_System.Core_test.dll", "binding_tests.dll" ];
+var assemblies = [ "mscorlib.dll", "System.dll", "System.Core.dll", "System.Net.Http.dll", "Mono.Security.dll", "main.exe", "nunitlite.dll", "mini_tests.dll", "wasm_corlib_test.dll", "wasm_System_test.dll", "wasm_System.Core_test.dll", "binding_tests.dll" ];
```
Comment (kumpera) [inline sdks/wasm/test.js]: UG, I need to take the time to replace this fugly piece of code to use runtime.g.js :(

File: mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs (hunk 2 above)
Comment (kumpera) [reply mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: Oh! Forgot one thing. We'll need to add GetHttpMessageHandler to the linker xml for System.Net.Http.
@akoeplinger can you help us with this?

Comment (kjpou1) [reply mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: How does the variable get set if it is private?

File: mcs/class/System.Net.Http/Makefile (hunk 1 above)
Comment (kjpou1) [reply mcs/class/System.Net.Http/Makefile]: Indeed it does work.

File: mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs (hunk 2 above)
Comment (kumpera) [reply mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: It would be set by reflection.

Comment (kjpou1) [reply mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: :-)

**Comment by kjpou1** [comment]
@monojenkins build OS X i386

File: mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs (hunk 2 above)
Comment (akoeplinger) [reply mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: you can add the `[PreserveDependency ("GetHttpMessageHandler", "System.Net.Http.HttpClient")]` attribute to GetDefaultHandler().

**Comment by kumpera** [comment]
@baulig have you read the associated issue? It covers your questions.

**Comment by kumpera** [comment]
@monojenkins build failed

**Comment by baulig** [comment]
Well, what I really do not like about this is that now several of our platforms each have their own mechanism for providing a default handler instead of this being handled in unified way.

File: mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs (hunk 2 above)
Comment (kjpou1) [reply mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: Thanks @akoeplinger

File: mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+using System;
```
Comment (marek-safar) [inline mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: Please use consistent spacing

Comment (kjpou1) [reply mcs/class/System.Net.Http/System.Net.Http/HttpClient.wasm.cs]: Cleaned up

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
Thanks @marek-safar

**Comment by masudanmurmu** [comment]
supported

**Pull request merged**